File: lua/hob/term.lua
-- --- hob.term ---
-- Terminal interaction for flows.
--
-- Everything here is an effect, so the engine records it, and a preview can
-- skip or redirect it, rather than a flow writing straight to the terminal.
-- Styling is named rather than spelled in escape codes: the engine owns the
-- terminal and decides whether colour makes sense.
--
-- Questions are written to stderr and answers are read from stdin, so stdout
-- stays the data a caller can pipe.

local hob = require("hob")

local term = {}

-- One table per function so a typo is an error here rather than a silently
-- ignored key that reaches the engine.
local ALLOWED = {
  print = { style = true, stream = true },
  input = { prompt = true, default = true, initial = true },
  allow = { prompt = true, default = true, detail = true },
  select = { prompt = true, options = true, default = true },
  choose = { prompt = true, options = true, defaults = true, min = true, max = true },
}

local function check(op, opts)
  for key in pairs(opts) do
    if not ALLOWED[op][key] then
      error("hob.term." .. op .. ": unknown option `" .. key .. "`", 0)
    end
  end
end

--- Print one line for the user.
--
-- `opts.style` is one of "error", "warn", "info", "success", "muted" or
-- "title". `opts.stream` is "stdout" (the default) or "stderr".
function term.print(text, opts)
  opts = opts or {}
  check("print", opts)
  return hob.effect("term", "print", {
    text = text,
    style = opts.style,
    stream = opts.stream,
  })
end

--- Ask for one line of input.
--
-- `opts.prompt` is the question, `opts.default` the answer used when the line
-- is empty. `opts.initial` is accepted as a second default; hob has no line
-- editor to pre-fill.
function term.input(opts)
  opts = opts or {}
  check("input", opts)
  return hob.effect("term", "input", {
    prompt = opts.prompt,
    default = opts.default,
    initial = opts.initial,
  })
end

--- Ask a yes/no question; an empty line takes `opts.default` (false without).
--
-- `opts.detail` is printed before the question, for the text a confirmation
-- should show: what is about to happen, not just "are you sure".
function term.allow(prompt, opts)
  opts = opts or {}
  check("allow", opts)
  return hob.effect("term", "allow", {
    prompt = prompt,
    default = opts.default,
    detail = opts.detail,
  })
end

--- Pick one option from a numbered list.
--
-- `opts.options` holds labels, or tables `{ label = ..., value = ...}` when the
-- label shown is not the value wanted. `opts.default` is the label used when
-- the line is empty.
function term.select(opts)
  opts = opts or {}
  check("select", opts)
  return hob.effect("term", "select", {
    prompt = opts.prompt,
    options = opts.options,
    default = opts.default,
  })
end

--- Pick several options from a numbered list.
--
-- `opts.defaults` are the labels used when the line is empty; `opts.min` and
-- `opts.max` bound how many picks are accepted.
function term.choose(opts)
  opts = opts or {}
  check("choose", opts)
  return hob.effect("term", "choose", {
    prompt = opts.prompt,
    options = opts.options,
    defaults = opts.defaults,
    min = opts.min,
    max = opts.max,
  })
end

return term

File: src/effect/ops/term.rs
// --- effect::ops::term ---
// Arguments of `term.print`.

use serde::Deserialize;

/// Named styling, so a flow never spells an escape code.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "lowercase")]
pub(crate) enum Style {
    /// Something failed.
    Error,
    /// Something recoverable went wrong.
    Warn,
    /// Neutral information.
    Info,
    /// Something worked.
    Success,
    /// Secondary text.
    Muted,
    /// A heading.
    Title,
}

impl Style {
    /// The ANSI SGR parameter for this style.
    pub(crate) fn code(self) -> &'static str {
        match self {
            Self::Error => "31",
            Self::Warn => "33",
            Self::Info => "34",
            Self::Success => "32",
            Self::Muted => "2",
            Self::Title => "1",
        }
    }
}

/// Which stream a line goes to.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "lowercase")]
pub(crate) enum Stream {
    /// Standard output, the default.
    Stdout,
    /// Standard error.
    Stderr,
}

/// One line of output.
#[derive(Debug, Deserialize)]
pub(crate) struct Print {
    /// The text to print.
    pub(crate) text: String,
    /// Optional styling.
    #[serde(default)]
    pub(crate) style: Option<Style>,
    /// Optional stream override.
    #[serde(default)]
    pub(crate) stream: Option<Stream>,
}

/// `term.input`.
#[derive(Debug, Deserialize)]
pub(crate) struct Input {
    /// Question shown before reading.
    #[serde(default)]
    pub(crate) prompt: Option<String>,
    /// Answer used when the line is empty.
    #[serde(default)]
    pub(crate) default: Option<String>,
    /// Accepted as a second default: hob has no line editor to pre-fill.
    #[serde(default)]
    pub(crate) initial: Option<String>,
}

/// `term.allow`.
#[derive(Debug, Deserialize)]
pub(crate) struct Allow {
    /// Question shown before reading.
    #[serde(default)]
    pub(crate) prompt: Option<String>,
    /// Answer used when the line is empty; false when absent.
    #[serde(default)]
    pub(crate) default: Option<bool>,
    /// Text printed before the question, e.g. what is about to happen.
    #[serde(default)]
    pub(crate) detail: Option<String>,
}

/// One selectable option: a bare label, or a label with a value to return.
#[derive(Debug, Deserialize)]
#[serde(untagged)]
pub(crate) enum Choice {
    /// The label is the value.
    Label(String),
    /// The label is shown, the value is returned.
    Pair {
        /// Text shown in the list.
        label: String,
        /// Value returned when this option is picked.
        value: serde_json::Value,
    },
}

impl Choice {
    /// Text shown in the list.
    pub(crate) fn label(&self) -> &str {
        match self {
            Self::Label(label) | Self::Pair { label, .. } => label,
        }
    }

    /// Value returned when this option is picked.
    pub(crate) fn value(&self) -> serde_json::Value {
        match self {
            Self::Label(label) => serde_json::Value::String(label.clone()),
            Self::Pair { value, .. } => value.clone(),
        }
    }
}

/// `term.select`.
#[derive(Debug, Deserialize)]
pub(crate) struct Select {
    /// Question shown above the list.
    #[serde(default)]
    pub(crate) prompt: Option<String>,
    /// Options to pick from.
    pub(crate) options: Vec<Choice>,
    /// Label picked when the line is empty.
    #[serde(default)]
    pub(crate) default: Option<String>,
}

/// `term.choose`.
#[derive(Debug, Deserialize)]
pub(crate) struct Choose {
    /// Question shown above the list.
    #[serde(default)]
    pub(crate) prompt: Option<String>,
    /// Options to pick from.
    pub(crate) options: Vec<Choice>,
    /// Labels picked when the line is empty.
    #[serde(default)]
    pub(crate) defaults: Vec<String>,
    /// Fewest picks accepted.
    #[serde(default)]
    pub(crate) min: Option<usize>,
    /// Most picks accepted.
    #[serde(default)]
    pub(crate) max: Option<usize>,
}
